gossip/client_test.go

```diff
@@ -105,3 +179,36 @@ func TestClientGossip(t *testing.T) {
 		return nil
 	})
 }
+
+// TestClientNodeID verifies a client's gossip request with correct NodeID.
+func TestClientNodeID(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	local, remote, stopper := startFakeServerGossip(t)
+	disconnected := make(chan *client, 1)
+
+	// Use an insecure context. We're talking to unix socket which are not in the certs.
+	lclock := hlc.NewClock(hlc.UnixNano)
+	rpcContext := rpc.NewContext(&base.Context{Insecure: true}, lclock, stopper)
+
+	// Start a gossip client.
+	c := newClient(remote.nodeAddr)
+	defer func() {
+		stopper.Stop()
+		if c != <-disconnected {
+			t.Errorf("expected client disconnect after remote close")
+		}
+	}()
+	c.start(local, disconnected, rpcContext, stopper)
+	// Wait for c.gossip to start.
+	<-remote.firstMsgChan
+	nodeID := roachpb.NodeID(1)
+	// Simulate a nodeID setting after c.gossip started.
+	local.SetNodeID(nodeID)
+	// Ignore the msg with old nodeID, Wait for the new nodeID to take effect.
+	time.Sleep(10 * time.Millisecond)
```

Comment:
Can you do this without the sleep? Maybe have the fake gossip server write to the channel after every message, not just the first?  If you must sleep, use `util.SucceedsWithin` to retry the test for up to 1 second. 10ms is often not enough on circleci.